### PR TITLE
chore(client, cli): implement --data-proxy [DPGA, 1]

### PR DESCRIPTION
**Pull request by millsp**
This PR replaces the `PRISMA_CLIENT_ENGINE_TYPE=dataproxy` mechanism in favor of `--data-proxy` via the cli.

Next https://github.com/prisma/prisma/pull/13512

https://github.com/orgs/prisma/projects/70#card-82299232

### CODE REVIEW

File: helpers/compile/build.ts [hunk 1]
```diff
@@ -193,8 +193,9 @@ const watch =
       }
     }, 10)
 
-    changeWatcher.on('change', onChange)
-    restartWatcher.once('add', onAdd)
+    changeWatcher.on('change', fastRebuild)
+    restartWatcher.once('add', fullRebuild)
+    restartWatcher.once('unlink', fullRebuild)
```
Comment (millsp) [inline helpers/compile/build.ts:198]: Unrelated changes, triggers a full rebuild on file deletion.

File: packages/cli/src/Generate.ts [hunk 2]
```diff
@@ -136,6 +138,7 @@ ${chalk.bold('Examples')}
         printDownloadProgress: !watchMode,
         version: enginesVersion,
         cliVersion: pkg.version,
+        dataProxy: args['--data-proxy'],
```
Comment (millsp) [inline packages/cli/src/Generate.ts:141]: Passing --data-proxy it to getGenerators, to pass it to the generator.

File: packages/cli/src/Generate.ts [hunk 3]
```diff
@@ -48,9 +48,10 @@ ${chalk.bold('Usage')}
 
 ${chalk.bold('Options')}
 
-  -h, --help   Display this help message
-    --schema   Custom path to your Prisma schema
-     --watch   Watch the Prisma schema and rerun after a change
+    -h, --help   Display this help message
+      --schema   Custom path to your Prisma schema
+  --data-proxy   Enable the Data Proxy in the Prisma Client
```
Comment (millsp) [inline packages/cli/src/Generate.ts:53]: New --data-proxy option to replace PRISMA_CLIENT_ENGINE_TYPE=dataproxy.

File: packages/client/src/__tests__/integration/happy/custom-engine-binary/test.ts [hunk 4]
```diff
@@ -5,11 +5,6 @@ import path from 'path'
 
 import { generateTestClient } from '../../../../utils/getTestClient'
 
-if (getClientEngineType() === ClientEngineType.DataProxy) {
```
Comment (millsp) [inline packages/client/src/__tests__/integration/happy/custom-engine-binary/test.ts:8]: Not needed.

File: packages/client/src/generation/utils/buildNFTAnnotations.ts [hunk 5]
```diff
@@ -16,10 +16,14 @@ import { map } from '../../../../../helpers/blaze/map'
  * @returns
  */
 export function buildNFTAnnotations(
```
Comment (danstarns) [inline packages/client/src/generation/utils/buildNFTAnnotations.ts:18]: The args for this function are getting lengthy, wdyt about making it an obj?

File: packages/client/src/generation/TSClient/TSClient.ts [hunk 6]
```diff
@@ -86,15 +88,16 @@ export class TSClient implements Generatable {
       engineVersion: this.options.engineVersion,
       datasourceNames: datasources.map((d) => d.name),
       activeProvider: this.options.activeProvider,
+      dataProxy: this.options.dataProxy,
```
Comment (millsp) [inline packages/client/src/generation/TSClient/TSClient.ts:91]: `dataProxy` gets written into the generated client config.

File: packages/client/src/generation/generateClient.ts [hunk 7]
```diff
@@ -146,13 +148,14 @@ export async function generateClient({
...
   activeProvider,
+  dataProxy,
 }: GenerateClientOptions): Promise<void> {
   const useDotPrisma = testMode ? !runtimeDir : !generator?.isCustomOutput
   const clientEngineType = getClientEngineType(generator!)
   runtimeDir = runtimeDir || (useDotPrisma ? '@prisma/client/runtime' : './runtime')
 
   // we make sure that we point to the right engine build
-  if (clientEngineType === ClientEngineType.DataProxy) {
+  if (dataProxy === true) {
```
Comment (millsp) [inline packages/client/src/generation/generateClient.ts:158]: We don't use `ClientEngineType.DataProxy` any longer.

File: packages/client/src/generation/generator.ts [hunk 8]
```diff
@@ -51,6 +51,7 @@ if (process.argv[1] === __filename) {
         clientVersion,
         transpile: true,
         activeProvider: options.datasources[0]?.activeProvider,
+        dataProxy: options.dataProxy,
```
Comment (millsp) [inline packages/client/src/generation/generator.ts:54]: Passing the option from cli => getGenerator => Generator => .setOptions()

File: packages/client/src/generation/utils/buildDirname.ts [hunk 9]
```diff
@@ -1,19 +1,18 @@
...
-export function buildDirname(clientEngineType: ClientEngineType, relativeOutdir: string, runtimeDir: string) {
-  if (clientEngineType !== ClientEngineType.DataProxy) {
-    return buildDirnameFind(relativeOutdir, runtimeDir)
+export function buildDirname(dataProxy: boolean | undefined, relativeOutdir: string, runtimeDir: string) {
+  if (dataProxy === true) {
+    return buildDirnameDefault()
   }
 
-  return buildDirnameDefault()
+  return buildDirnameFind(relativeOutdir, runtimeDir)
```
Comment (millsp) [inline packages/client/src/generation/utils/buildDirname.ts:15]: Just swapping the logic around.

File: packages/client/src/runtime/getPrismaClient.ts [hunk 10]
```diff
@@ -448,8 +457,11 @@ export function getPrismaClient(config: GetPrismaClientConfig) {
     get [Symbol.toStringTag]() {
       return 'PrismaClient'
     }
+
     private getEngine(): Engine {
-      if (this._clientEngineType === ClientEngineType.Library) {
+      if (this._dataProxy === true) {
```
Comment (millsp) [inline packages/client/src/runtime/getPrismaClient.ts:462]: We now control the engine selection via `--data-proxy`.

File: packages/sdk/src/client/getClientEngineType.ts [hunk 11]
```diff
@@ -3,7 +3,6 @@ import type { GeneratorConfig } from '@prisma/generator-helper'
 export enum ClientEngineType {
   Library = 'library',
   Binary = 'binary',
-  DataProxy = 'dataproxy',
```
Comment (millsp) [inline packages/sdk/src/client/getClientEngineType.ts:6]: This hack is not a valid one anymore.

File: packages/client/src/__tests__/buildNFTAnnotations.test.ts [hunk 12]
```diff
@@ -84,16 +86,15 @@ describe('binary', () => {
 describe('dataproxy', () => {
   it('generates no annotations', () => {
     const annotations = buildNFTAnnotations(
-      ClientEngineType.DataProxy,
+      true,
```
Comment (danstarns) [inline packages/client/src/__tests__/buildNFTAnnotations.test.ts:89]: It is difficult to tell what this boolean does without looking at the func signature.

File: packages/client/src/generation/utils/buildNFTAnnotations.ts [hunk 13]
```diff
@@ -16,10 +16,14 @@ import { map } from '../../../../../helpers/blaze/map'
...
+  dataProxy: boolean | undefined,
```
[the remark this answers is not in the document]
Comment (millsp) [reply packages/client/src/generation/utils/buildNFTAnnotations.ts:19]: Yeah definitely and it was originally written this way, I haven been banging my head for hours on this. I initially had it as a boolean, not an optional boolean. I was facing an odd problem where making them required caused the CI to hang and never finish the tests, although it perfectly passes locally. So I'm leaving it like this for now as I am running short on time.

File: packages/client/src/generation/utils/buildNFTAnnotations.ts (hunk 5 above)
Comment (aqrln) [reply packages/client/src/generation/utils/buildNFTAnnotations.ts:18]: +1

Since the PR needs to be merged ASAP, it's fine to change this later, but please post it to `#project-ts-orm-tech-debt` so it's not lost.

File: packages/client/src/generation/generateClient.ts [hunk 14]
```diff
@@ -57,7 +58,6 @@ export interface BuildClientResult {
 
 // eslint-disable-next-line @typescript-eslint/require-await
 export async function buildClient({
-  datamodel,
```
Comment (janpio) [inline packages/client/src/generation/generateClient.ts:60]: Unrelated change?

File: packages/client/src/generation/utils/buildDMMF.ts [hunk 15]
```diff
@@ -1,15 +1,14 @@
...
  * Creates the necessary declarations to embed the generated DMMF into the
  * generated client. It compresses the DMMF for the data proxy engine.
- * @param engineType
+ * @param dataProxy
  * @param dmmf
  * @returns
  */
-export function buildDMMF(engineType: ClientEngineType, dmmf: string) {
-  if (engineType === ClientEngineType.DataProxy) {
+export function buildDMMF(dataProxy: boolean | undefined, dmmf: string) {
```
Comment (janpio) [inline packages/client/src/generation/utils/buildDMMF.ts:10]: Why accept `undefined`?

Comment (janpio) [reply packages/client/src/generation/utils/buildDMMF.ts:10]: Ah found https://github.com/prisma/prisma/pull/13561/files#r885447162 further down

File: packages/engine-core/src/data-proxy/DataProxyEngine.ts [hunk 16]
```diff
@@ -237,12 +237,9 @@ export class DataProxyEngine extends Engine {
...
     if (protocol !== 'prisma:') {
-      throw new InvalidDatasourceError(
-        'Datasource URL should use prisma:// protocol. If you are not using the Data Proxy, remove the `dataProxy` from the `previewFeatures` in your schema and ensure that `PRISMA_CLIENT_ENGINE_TYPE` environment variable is not set to `dataproxy`.',
-        {
-          clientVersion: this.clientVersion,
-        },
-      )
+      throw new InvalidDatasourceError('Datasource URL must use prisma:// protocol when --data-proxy is used', {
+        clientVersion: this.clientVersion,
+      })
```
Comment (janpio) [inline packages/engine-core/src/data-proxy/DataProxyEngine.ts:242]: Does this have a test? Did not see any when scrolling through the PR.

**Review by janpio — APPROVED** [review]
Scroll through review: All changes make sense and nothing pops out as problematic.

File: packages/client/src/generation/utils/buildNFTAnnotations.ts (hunk 13 above)
[the remark this answers is not in the document]
Comment (millsp) [reply packages/client/src/generation/utils/buildNFTAnnotations.ts:19]: Turned out that it is coming from `ts-jest` https://github.com/kulshekhar/ts-jest/issues/3507, fixed by passing `isolatedModules: true`.

File: packages/client/src/generation/generateClient.ts (hunk 14 above)
Comment (millsp) [reply packages/client/src/generation/generateClient.ts:60]: Yep, unsused.

File: packages/client/src/generation/utils/buildNFTAnnotations.ts (hunk 5 above)
Comment (millsp) [reply packages/client/src/generation/utils/buildNFTAnnotations.ts:18]: https://www.notion.so/prismaio/Technical-Debt-TS-ORM-c199f16035304e2db13793f176098c24?p=78d58e9a82064605890491722be2f0ff

File: packages/client/src/__tests__/buildNFTAnnotations.test.ts (hunk 12 above)
Comment (millsp) [reply packages/client/src/__tests__/buildNFTAnnotations.test.ts:89]: https://www.notion.so/prismaio/Technical-Debt-TS-ORM-c199f16035304e2db13793f176098c24?p=78d58e9a82064605890491722be2f0ff

File: packages/client/src/generation/utils/buildNFTAnnotations.ts [hunk 17]
```diff
@@ -95,7 +99,9 @@ function buildNFTEngineAnnotation(engineType: ClientEngineType, platform: Platfo
  * @returns
  */
 function buildNFTSchemaAnnotation(engineType: ClientEngineType, relativeOutdir: string) {
-  if (engineType === ClientEngineType.DataProxy) return ''
+  if (engineType === ClientEngineType.Library || engineType === ClientEngineType.Binary) {
+    return buildNFTAnnotation('schema.prisma', relativeOutdir)
+  }
 
-  return buildNFTAnnotation('schema.prisma', relativeOutdir)
+  return undefined
```
[the remark this answers is not in the document]
Comment (millsp) [reply packages/client/src/generation/utils/buildNFTAnnotations.ts:106]: Nice catch, thanks

File: packages/engine-core/src/data-proxy/DataProxyEngine.ts (hunk 16 above)
Comment (millsp) [reply packages/engine-core/src/data-proxy/DataProxyEngine.ts:243]: Note taken, test will be added in finalization PR.

File: packages/client/src/generation/utils/buildNFTAnnotations.ts [hunk 18]
```diff
@@ -95,7 +99,9 @@ function buildNFTEngineAnnotation(engineType: ClientEngineType, platform: Platfo
  * @returns
  */
 function buildNFTSchemaAnnotation(engineType: ClientEngineType, relativeOutdir: string) {
-  if (engineType === ClientEngineType.DataProxy) return ''
+  if (engineType === ClientEngineType.Library || engineType === ClientEngineType.Binary) {
```
Comment (aqrln) [inline packages/client/src/generation/utils/buildNFTAnnotations.ts:102]: Isn't this condition always true now?

Comment (millsp) [reply packages/client/src/generation/utils/buildNFTAnnotations.ts:102]: Yes, it is. Thanks for that. I will fix it in the finalization PR.

Comment (millsp) [reply packages/client/src/generation/utils/buildNFTAnnotations.ts:102]: (although not critical since nothing would be found)

**Review by aqrln — APPROVED** [review]

File: packages/engine-core/src/data-proxy/utils/request.ts [hunk 19]
```diff
@@ -1,34 +1,40 @@
...
-export type RequestOptions = O.Patch<
-  {
-    headers?: { [k: string]: string }
-    body?: string
-  },
-  RequestInit
->
-
-// our implementation handles less
+export type RequestOptions = O.Patch<{ headers?: { [k: string]: string }; body?: string }, RequestInit>
```
Comment (jkomyno) [inline packages/engine-core/src/data-proxy/utils/request.ts:10]: Note: it seems this change prevents TypeScript to properly build `engine-core`, as it affects the `buildOptions` function.

Temporary solution:

```typescript
function buildOptions(options: RequestOptions): Https.RequestOptions {
  return {
    // @ts-ignore
    method: options.method,
    headers: buildHeaders(options),
  }
}
```
